rust/nasl-builtin-cert/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+    ///
+    /// - not-after  The notAfter time as UTC value in ISO time format
+    ///                 (e.g. "20280929T143520").
+    ///
+    /// - all Return all available information in a human readable
+    ///          format.  Not yet implemented.
+    ///
+    /// - hostnames Return an array with all hostnames listed in the
+    ///   certificates, i.e. the CN part of the subject and all dns-name
+    ///   type subjectAltNames.
+    ///
+    /// - fpr-sha-256 The SHA-256 fingerprint of the certificate.  The
+    ///                  fingerprint is, as usual, computed over the entire
+    ///                  DER encode certificate.
+    ///
+    /// - fpr-sha-1   The SHA-1 fingerprint of the certificate.  The
+    ///                  fingerprint is, as usual, computed over the entire
+    ///                  DER encode certificate.
+    ///
+    /// - image       Return the entire certificate as binary data.
+    ///
+    /// - algorithm-name  Same as signature-algorithm-name. TODO: Remove it and
+    ///                      leave only signature-algorithm-name.
+    ///
+    /// - signature-algorithm-name  Return the algorithm name used to sign the
+    ///                                certificate. Get the OID of the digest
+    ///                                algorithm and translated to a name from a
+    ///                                list from Wireshark.
+    ///                                See epan/dissectors/packet-pkcs1.c
+    ///
+    /// - public-key-algorithm-name  Return the algorithm name of the public key.
+    ///
+    /// - modulus      Return the RSA public key's modulus found in the
+    ///                   structure of the given cert.
+    ///
+    /// - exponent    Return the RSA public key's exponent found in
+    ///                  the structure of the given cert.
+    ///
+    /// - key-size    Return the size to hold the parameters size in bits.
+    ///                  For RSA the bits returned is the modulus.
+    ///                  For DSA the bits returned are of the public exponent.
+    ///
+    ///
+    /// The following arguments are required:
+    /// - pos(0): Object id of the certificate.
+    ///
+    /// - pos(1): A string with the command to select what to return; see above.
+    ///
+    /// The following arguments are optional:
+    /// - idx Used by certain commands to select the n-th value of a set
+    ///    of values.  If not given 0 is assumed.
+    ///
+    /// A NASL type depending on the used command.
+    #[nasl_function(named(idx))]
+    fn cert_query(
+        &self,
+        fd: usize,
+        query: &str,
+        idx: Option<usize>,
+    ) -> Result<NaslValue, FunctionErrorKind> {
+        let idx = idx.unwrap_or(0);
+        let handle = self.0.read().unwrap();
+        if fd >= handle.certs.len() {
+            return Err(FunctionErrorKind::WrongArgument(
+                "The given file descriptor is not valid.".to_string(),
+            ));
```

Comment:
Technically not, but we always used the term `fd` for everything in the c NASL code, that had underlying data stored internally. In this case the underlying data is the certificate, which is stored. For example, in the socket implementation, the `fd` is also an internal handle. The real `fd` given by the system is stored within the rust socket implementation.
A more appropriate name would be an index or something like that, but as it is widely used within NASL, it would be hard to change.